cdc/processor/sinkmanager/worker_impl.go

```diff
@@ -221,3 +240,115 @@ func (w *workerImpl) advanceTableSink(t *tableSinkTask, commitTs model.Ts, size
 	w.memQuota.record(t.tableID, resolvedTs, size)
 	return t.tableSink.updateResolvedTs(resolvedTs)
 }
+
+type redoWorkerImpl struct {
+	changefeedID   model.ChangeFeedID
+	sortEngine     sorter.EventSortEngine
+	memQuota       *memQuota
+	redoManager    redo.LogManager
+	eventCache     *redoEventCache
+	splitTxn       bool
+	enableOldValue bool
+}
+
+func newRedoWorker(
+	changefeedID model.ChangeFeedID,
+	sortEngine sorter.EventSortEngine,
+	quota *memQuota,
+	redoManager redo.LogManager,
+	eventCache *redoEventCache,
+	splitTxn bool,
+	enableOldValue bool,
+) redoWorker {
+	return &redoWorkerImpl{
+		changefeedID:   changefeedID,
+		sortEngine:     sortEngine,
+		memQuota:       quota,
+		redoManager:    redoManager,
+		eventCache:     eventCache,
+		splitTxn:       splitTxn,
+		enableOldValue: enableOldValue,
+	}
+}
+
+func (w *redoWorkerImpl) handleTasks(ctx context.Context, taskChan <-chan *redoTask) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case task := <-taskChan:
+			rows := make([]*model.RowChangedEvent, 0, 1024)
+			batchSize := uint64(0)
+			cache := w.eventCache.getAppender(task.tableID)
+			cachedSize := uint64(0)
+			memAllocated := true
+
+			var lastPos sorter.Position
+			var maybeEmitBatchEvents = func(allFinished, txnFinished bool) error {
+				if batchSize == 0 || (!allFinished && batchSize < requestMemSize) {
+					return nil
+				}
+
+				releaseMem := func() { w.memQuota.refund(batchSize - cachedSize) }
+				err := w.redoManager.EmitRowChangedEvents(ctx, task.tableID, releaseMem, rows...)
+				if err != nil {
+					return errors.Trace(err)
+				}
+				if lastPos.Valid() {
+					err = w.redoManager.UpdateResolvedTs(ctx, task.tableID, lastPos.CommitTs)
+					if err != nil {
+						return errors.Trace(err)
+					}
+				}
+
+				rows = rows[0:]
+				if cap(rows) > 1024 {
+					rows = make([]*model.RowChangedEvent, 0, 1024)
+				}
+				batchSize = 0
+				cachedSize = 0
+
+				if !allFinished {
+					if !txnFinished {
+						w.memQuota.forceAcquire(requestMemSize)
+					} else {
+						memAllocated = w.memQuota.tryAcquire(requestMemSize)
+					}
+				}
+				return nil
+			}
+
+			// lowerBound and upperBound are both closed intervals.
+			iter := w.sortEngine.FetchByTable(task.tableID, task.lowerBound, task.getUpperBound())
+			defer iter.Close()
+			for memAllocated {
+				e, pos, err := iter.Next()
+				if err != nil {
+					return errors.Trace(err)
+				}
+				if e == nil {
+					// There is no more data.
+					maybeEmitBatchEvents(true, true)
+					return nil
+				}
+				if pos.Valid() {
+					lastPos = pos
+				}
+
+				x, size, err := convertRowChangedEvents(w.changefeedID, task.tableID, w.enableOldValue, e)
+				if err != nil {
+					return errors.Trace(err)
+				}
+
+				rows = append(rows, x...)
+				batchSize += size
+				if cache.push(x[0], size, pos.Valid()) {
```

Comment:
Why only the first event?

Reply:
Because the slice length will always be 1.

Reply:
Nope. It can be 2. `				rowChangedEvents = append(rowChangedEvents, deleteEvent.Row, insertEvent.Row)
`

Reply:
So nice catch. Thank you!